bitcoin/src/pow.rs

```diff
@@ -248,14 +257,31 @@ impl Target {
     ///
     /// The difficulty can only decrease or increase by a factor of 4 max on each difficulty
     /// adjustment period.
-    pub fn min_difficulty_transition_threshold(&self) -> Self { Self(self.0 >> 2) }
+    pub fn min_difficulty_transition_threshold(&self) -> Self {
+        let min = self.0 >> 2;   // May shift to zero but we want
+        Self(min.max(U256::ONE)) // difficulty transition to be at least 1
```

Comment:
In 8f3537b17714db104c4dccdfc8614e09a76fc381:

As methods on *Target* these are confusingly named. This function is called `min_difficulty` but it actually computes the minimum *target*, which is the maximum difficulty.

Also, I am digging into the Core source code and can't see any maximum difficulty like this... it's completely academic because there is no way the target can ever get close to 1 but we ought to be correct nonetheless and just let the value go to 0.

Reply:
> As methods on Target these are confusingly named. This function is called min_difficulty but it actually computes the minimum target, which is the maximum difficulty.

Renamed and deprecated the originals.

> Also, I am digging into the Core source code and can't see any maximum difficulty like this... it's completely academic because there is no way the target can ever get close to 1 but we ought to be correct nonetheless and just let the value go to 0.

Now I'm confused, previously above you said "We also need to check that our minimum shift doesn't go below 1. (And unlike the maximum, the minimum limit has actually been hit on the mainchain many times.)"

Reply:
Yes, there is a minimum difficulty of 1. There is no minimum target.

Reply:
So you are saying that plain old `Self(self.0 >> 2)` is fine for `min_transition_threshold`, right? And since its a threshold folk should not be calling `difficulty()` on the returned target anyways (which would be div by zero and panic).

Reply:
> So you are saying that plain old `Self(self.0 >> 2)` is fine for `min_transition_threshold`, right?

Correct. This matches what Bitcoin Core does. We should add a comment I guess, highlighting this.

>And since its a threshold folk should not be calling `difficulty()` on the returned target anyways (which would be div by zero and panic).

Hmmm, I'd rather not get a div-by-0 panic even though this branch is literally impossible to hit on a real chain (unless sha2 is broken, in which case I've probably moved to a new town with a fake mustache and you won't be hearing any cryptography from me again..).

Maybe in `difficulty()` we should special-case the div by 0 and return the maximum value, or zero, or something. I don't think it matters what we return because this case is impossible on-chain, but we should pick something and document it. (My vote is to return whatever the maximum difficulty value is, so that at least comparison operations will do the most logical thing.)

Reply:
If we're going to not panic then definitely max. But I guess panic is easier to debug. Maybe it could be `debug_assert`.

Reply:
I don't think we need to make this "easier to debug" ... my feeling is that this is a "garbage in garbage out" case and if somebody is actually calling `difficulty` on a zero target, they would prefer their program output something nonsensical rather than panic. (My thinking is: if they were doing something "real" then this case would be impossible; conversely, if it's possible then they're implementing some sort of toy or calculator or something.)

But I could be convinced either way.

Reply:
It could be a mistake if they somehow accidentally produced 0. Things like that happen all the time. `debug_assert` at least makes it possible to find it more quickly.

Reply:
Ok, sure, I'm convinced. Let's add a debug_assert here.